Pyomo doe refactor

djlaky opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

## Fixes ##
Pyomo.DoE interface was made more intuitive and some bugs were removed.

## Summary/Motivation:
Pyomo.DoE was refactored to coincide with the new "experiment-based" structure of ParmEst. In this way, the packages should overlap with syntax and workflows using both tools should be easier to implement. Also, the existing interface was clunky and required a few bugfixes.

## Changes proposed in this PR:
- Overhaul of the Pyomo.DoE code and Class/Object structure

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

mrmundt commented:
@djlaky - There will be a failure in the `singletest` suite that is not related, so please ignore it.

blnicho commented:
@djlaky @adowling2 I'm guessing this PR supersedes most/all of the changes in #3250. Can that PR be closed?

djlaky commented:
> @djlaky @adowling2 I'm guessing this PR supersedes most/all of the changes in #3250. Can that PR be closed?

Yes, this will supersede those changes.

mrmundt commented:
@djlaky - The pypy failure appears to actually be real, so please go look into it.

mrmundt commented:
@djlaky - There may be a failure because of the BARON download that is unrelated. I have a PR open for it now.

mrmundt reviewed (changes requested):
General comment: for all new files, please add the copyright statement to the top:

```
#  ___________________________________________________________________________
#
#  Pyomo: Python Optimization Modeling Objects
#  Copyright (c) 2008-2024
#  National Technology and Engineering Solutions of Sandia, LLC
#  Under the terms of Contract DE-NA0003525 with National Technology and
#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
#  rights in this software.
#  This software is distributed under the 3-clause BSD License.
#  ___________________________________________________________________________
```

djlaky commented:
@blnicho I addressed most of the comments made in the group call on Friday, August 2nd. The code we went over (the generate scenarios function and the constructor) should be ready for re-review when you get the chance.

blnicho commented on pyomo/contrib/doe/examples/reactor_compute_factorial_FIM.py:
```diff
@@ -0,0 +1,98 @@
+    DATA_DIR = Path(__file__).parent
+    file_path = DATA_DIR / "result.json"
+
+    f = open(file_path)
+    data_ex = json.load(f)
```
You should use a context manager when working with files to ensure that the file always gets closed

blnicho commented on pyomo/contrib/doe/examples/reactor_compute_factorial_FIM.py:
```diff
@@ -0,0 +1,98 @@
+    scale_nominal_param_value = True
+
+    # Create the DesignOfExperiments object
+    # We will not be passing any prior information in this example
+    # and allow the experiment object and the DesignOfExperiments
+    # call of ``run_doe`` perform model initialization.
```
This comment is a little confusing as I was expecting to see a call to `run_doe` here. Maybe rephrase this as you're "relying on the model initialization routine built into the DesignOfExperiments class"

blnicho commented on pyomo/contrib/doe/examples/reactor_compute_factorial_FIM.py:
```diff
@@ -0,0 +1,98 @@
+        _only_compute_fim_lower=True,
+    )
+
+    # Make design ranges to compute the full factorial design
+    design_ranges = {"CA[0]": [1, 5, 9], "T[0]": [300, 700, 9]}
```
Maybe add a comment with a hint at the expected input format for this dictionary. It took me some time to figure out that the third number in the list was the number of steps.

`{<var_name>: [start, stop, numsteps]}`

blnicho commented on pyomo/contrib/doe/examples/reactor_compute_factorial_FIM.py:
```diff
@@ -0,0 +1,98 @@
+
+    # Plot the results
+    doe_obj.draw_factorial_figure(
+        sensitivity_design_variables=["CA[0]", "T[0]"],
+        fixed_design_variables={
+            "T[0.125]": 300,
+            "T[0.25]": 300,
+            "T[0.375]": 300,
+            "T[0.5]": 300,
+            "T[0.625]": 300,
+            "T[0.75]": 300,
+            "T[0.875]": 300,
+            "T[1]": 300,
```
The fact that these inputs (and the data for `design_ranges` above) are still relying on string names seems like a flaw in the redesign.

djlaky replied:
@blnicho Do you have a suggestion for doing this? We want to allow the user to pass a dictionary-like data structure to fix certain design variables and vary others.

I don't think it is natural to add a suffix to the model to do this... Would it make most sense for the user to provide two lists, one of the model objects, the other of the fixed value (or ranges) and assume the ordering is consistent?

blnicho reviewed (commented):
@djlaky I'm 1/3 of the way through the files. Here are my comments so far.

blnicho commented on pyomo/contrib/doe/utils.py:
```diff
@@ -0,0 +1,97 @@
+    return scaled_FIM
+
+
+def get_parameters_from_suffix(suffix, fix_vars=False):
```
Is this utility used anywhere? I couldn't find any references to it in this PR.

djlaky replied:
This was going to be used with parmest's ``param_to_var`` functionality but since the method is still string based. This will be removed and added to the immediate "To-Do's".

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -0,0 +1,598 @@
+)
+import pyomo.common.unittest as unittest
+
+from pyomo.contrib.doe import *
+from pyomo.contrib.doe.tests.experiment_class_example import *
+from pyomo.contrib.doe.tests.experiment_class_example_flags import (
+    FullReactorExperimentBad,
+)
+from pyomo.contrib.doe.utils import *
```
Please use explicit imports instead of `import *`.

blnicho reviewed (commented):
@djlaky here are my next batch of comments. I have 4 more files to look at.

blnicho commented:
@djlaky the conflicts are coming from #3335 getting merged this morning, they should be pretty easy to resolve but let me know if you need help.

I'm fine with leaving the TODO you mentioned. I think this is getting close and I'm working finishing up my review.

djlaky commented:
> @djlaky you deleted some of the duplicate files that I noted in my previous review and then reverted those commits. Why is that?
> 
> ![image](https://private-user-images.githubusercontent.com/18425615/357184118-c8d7aa61-b25c-4a50-ba41-82d1cbc683fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0OTMwMzYsIm5iZiI6MTcyMzQ5MjczNiwicGF0aCI6Ii8xODQyNTYxNS8zNTcxODQxMTgtYzhkN2FhNjEtYjI1Yy00YTUwLWJhNDEtODJkMWNiYzY4M2ZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEyVDE5NTg1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0ODMwMzk0OWYyODFkNTY2YzExNzMyYzVlNTM4ZjQ4M2ZkN2E4NGYwMTg1Y2VjYWZiYzFiY2E3ZjkyNWY3NGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3A-1NuvhpXAKVVGT9WUcDt7dhjvXwePlV3Z6ClogHlk)

I was trying to resolve the conflict because I thought I broke something. I’ll take them out when I resolve it. Thanks for the reminder.

djlaky commented:
@blnicho I went ahead and deleted those files you mentioned. Please let me know if any were missed when I went through! I should have resolved the merge conflict.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
+        # Generate blocks for finite difference scenarios
+        def build_block_scenarios(b, s):
+            # Generate model for the finite difference scenario
+            b.transfer_attributes_from(model.base_model.clone())
```
This construction rule is referencing a model object outside of its scope which can be dangerous. You can access the model containing the block data object using `m = b.model()`.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
-                    + m.priorFIM[p, q] * self.fim_scale_constant_value
-                )
+                block_design_var = pyo.ComponentUID(
+                    d, context=model.scenario_blocks[0]
+                ).find_component_on(model.scenario_blocks[s])
```
Use `m` instead of `model` in this rule.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
+            # If the row is greater than or equal to the column, we are in the
+            # lower triangle region of the L and FIM matrices.
+            # This region is where our equations are well-defined.
+            if list(model.parameter_names).index(c) >= list(
+                model.parameter_names
+            ).index(d):
```
Using a model object in this rule coming from outside of this function scope. You should use `m` here but remember that you are adding this constraint on a block, `model.obj_cons` so the block is what will be passed in as the first argument.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1283,14 +1197,14 @@ def trace_calc(m):
             Use permutation() to get permutations, sgn() to get signature
             """
-            r_list = list(range(len(m.regression_parameters)))
+            r_list = list(range(len(model.parameter_names)))
```
Using a model object in this rule coming from outside of this function scope. You should use `m` here but remember that you are adding this constraint on a block, `model.obj_cons` so the block is what will be passed in as the first argument.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1283,14 +1197,14 @@ def trace_calc(m):
             Calculate FIM elements. Can scale each element with 1000 for performance
             """
-            return m.trace == sum(m.fim[j, j] for j in m.regression_parameters)
+            return model.trace == sum(model.fim[j, j] for j in model.parameter_names)
```
Same comment about using `model` instead of `m` in this rule.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1302,133 +1216,1015 @@ def det_general(m):
+            model.objective = pyo.Objective(expr=0)
+
+    # Check to see if the model has all the required suffixes
+    def check_model_labels(self, model=None):
+        """
+        Checks if the model contains the necessary suffixes for the
+        DoE model to be constructed automatically.
+
+        Parameters
+        ----------
+        model: model for suffix checking, Default: None, (self.model)
+
+        """
+        if model is None:
+            model = self.model.base_model
```
I don't think you need a default here since you always pass in a model every time you use it. Also the docstring didn't match the implementation. The docstring said it would use `self.model` as the default but the implementation was using `self.model.base_model`. In the future, this might be better as a utility function rather than a class method. 
```suggestion
    def check_model_labels(self, model):
        """
        Checks if the model contains the necessary suffixes for the
        DoE model to be constructed automatically.

        Parameters
        ----------
        model: model for suffix checking

        """
```

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1302,133 +1216,1015 @@ def det_general(m):
+
+        Parameters
+        ----------
+        model: model to perform the full factorial exploration on
+        design_ranges: dict of lists, of the form {<var_name>: [start, stop, numsteps]}
+        method: string to specify which method should be used
+                options are ``kaug`` and ``sequential``
```
The docstring doesn't match the implementation. `model` isn't in the argument list.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1302,133 +1216,1015 @@ def det_general(m):
+                "Design ranges keys must be a subset of experimental design names."
+            )
+
+        # ToDo: Add more objetive types? i.e., modified-E; G-opt; V-opt; etc?
```
```suggestion
        # ToDo: Add more objective types? i.e., modified-E; G-opt; V-opt; etc?
```

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1302,133 +1216,1015 @@ def det_general(m):
-        design_val: design variable values dict
-        fix_opt: if True, fix. Else, unfix
-        optimize: a dictionary, keys are design variable name, values are True or False, deciding if this design variable is optimized as DOF this time
+        title_text: name of the figure, a string
+        xlabel_text: x label title, a string.
+            In a 1D sensitivity curve, it is the design variable by which the curve is drawn.
+        font_axes: axes label font size
+        font_tick: tick label font size
+        log_scale: if True, the result matrix will be scaled by log10
```
The docstring doesn't match the implementation. `figure_file_name` is missing from the docstring.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1302,133 +1216,1015 @@ def det_general(m):
+
+        Parameters
+        ----------
+        title_text: name of the figure, a string
+        xlabel_text: x label title, a string.
+            In a 2D heatmap, it should be the second design variable in the design_ranges
+        ylabel_text: y label title, a string.
+            In a 2D heatmap, it should be the first design variable in the dv_ranges
+        font_axes: axes label font size
+        font_tick: tick label font size
+        log_scale: if True, the result matrix will be scaled by log10
```
`figure_file_name` missing from the docstring

djlaky commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
-            all_jacobian_info=jac_square,
-            prior_FIM=self.prior_FIM,
-            scale_constant_value=self.scale_constant_value,
+            doe_block = pyo.Block()
+            doe_block_name = unique_component_name(model, "design_of_experiments_block")
+            model.add_component(doe_block_name, doe_block)
```
I'll move this to the TODO list, thanks.

djlaky commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
+            comp.fix()
+
+        model.dummy_obj = pyo.Objective(expr=0, sense=pyo.minimize)
+        self.solver.solve(model, tee=self.tee)
```
I've added it as a TODO, thanks!

djlaky commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
 
-        return FIM_analysis
+        # Compute and record FIM
+        self.seq_FIM = self.seq_jac.T @ cov_y @ self.seq_jac + self.prior_FIM
```
The @ is for matrix multiplication in numpy as of python version 3.5

If we want to support earlier versions I can change the syntax to use numpy.matmul

blnicho replied:
No need to change anything. We only officially support Python 3.8+. I just wasn't sure how to easily search for that syntax so I figured I would ask.

djlaky commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
-
-        # generate parameter name list and value dictionary with index
-        var_name = list(self.param.keys())
+        if not hasattr(model, "objective"):
+            model.objective = pyo.Objective(expr=0, sense=pyo.minimize)
```
This is a good point. I will add a TODO on the compute FIM method to make sure the model does not have an objective function, or to deactivate the objective function if it has one. I think this isn't extremely important for this PR as we are addressing user-input models at a later date (as mentioned in a couple other comments.

djlaky commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
         # loop over measurement variables and their time points
-        for mname in self.measure_name:
+        for k, v in model.experiment_outputs.items():
+            name = k.name
             try:
-                kaug_no = col.index(mname)
+                kaug_no = col.index(name)
```
The `get_dsdp` method from pyomo.contrib.sensitivity_toolbox.sens uses string names. So we would have to change that upstream code to fix this as well. I can add a TODO if we have time later (and if we do more analysis with things like kaug) to update this code to avoid strings, but I do not know who else or what else depends on the get_dsdp method.

djlaky commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1255 +25,1169 @@
+                assert res.solver.termination_condition == "optimal"
+            except:
+                raise RuntimeError(
+                    "Model from experiment did not solve appropriately. Make sure the model is well-posed."
+                )
```
Added a TODO to make the error more verbose.

djlaky commented:
@blnicho I addressed (almost) all of your new reviews from last night. I left the few that needed some clarification unresolved (although they are "addressed" with a TODO or some other edit in the code).

djlaky commented:
> @djlaky could you add the TODO items we've talked about as additional items in #2610? (or close #2610 and open a new issue consolidating the TODO's from this PR and lingering items in the old issue)

@blnicho I made a new issue (#3345), so we can close #2610. Thanks for the suggestion!

mrmundt commented on .vs/CMakeWorkspaceSettings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
I have a feeling that this is something that VSCode made and isn't related to this PR. Please remove.

mrmundt commented on doc/OnlineDocs/contributed_packages/doe/doe.rst:
```diff
@@ -203,89 +153,87 @@ The goal of MBDoE is to optimize the experiment design variables :math:`\boldsym
-    :pyobject: create_model
+.. literalinclude:: ../../../../pyomo/contrib/doe/examples/reactor_experiment.py
+    :start-after: Create flexible model without data
+    :end-before: End equation def'n
```
```suggestion
    :end-before: End equation definition
```

mrmundt commented on doc/OnlineDocs/contributed_packages/doe/doe.rst:
```diff
@@ -203,89 +153,87 @@ The goal of MBDoE is to optimize the experiment design variables :math:`\boldsym
+Here we add data to the model and finalize the discretization. This step is required before the model can be labeled.
 
+.. literalinclude:: ../../../../pyomo/contrib/doe/examples/reactor_experiment.py
+    :start-after: End equation def'n
```
```suggestion
    :start-after: End equation definition
```

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -132,19 +132,17 @@ def test_reactor_fd_central_solve(self):
 
-        # Assert model solves
-        assert doe_obj.results["Solver Status"] == "ok"
+        # self.assertTrue(model solves
```
Looks like this got changed by mistake with a find-and-replace

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -132,19 +132,17 @@ def test_reactor_fd_central_solve(self):
+        self.assertTrue(doe_obj.results["Solver Status"] == "ok")
 
-        # Assert that Q, F, and L are the same.
+        # self.assertTrue(that Q, F, and L are the same.
```
Looks like this got changed by mistake with a find-and-replace

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -157,18 +155,16 @@ def test_reactor_fd_forward_solve(self):
+        self.assertTrue(doe_obj.results["Solver Status"] == "ok")
 
-        # Assert that Q, F, and L are the same.
+        # self.assertTrue(that Q, F, and L are the same.
```
Looks like this got changed by mistake with a find-and-replace

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -181,20 +177,16 @@ def test_reactor_fd_backward_solve(self):
+        self.assertTrue(doe_obj.results["Solver Status"] == "ok")
 
-        # Assert that Q, F, and L are the same.
+        # self.assertTrue(that Q, F, and L are the same.
```
Looks like this got changed by mistake with a find-and-replace

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -228,19 +218,17 @@ def test_reactor_obj_cholesky_solve(self):
+        self.assertTrue(doe_obj.results["Solver Status"] == "ok")
 
-        # Assert that Q, F, and L are the same.
+        # self.assertTrue(that Q, F, and L are the same.
```
Looks like this got changed by mistake with a find-and-replace

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -322,17 +302,16 @@ def test_reactor_grid_search(self):
-        # Assert length is correct
-        assert (len(CA_vals) == 9) and (len(T_vals) == 9)
-        assert (len(set(CA_vals)) == 3) and (len(set(T_vals)) == 3)
+        # self.assertTrue(length is correct
```
Looks like this got changed by mistake with a find-and-replace

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -322,17 +302,16 @@ def test_reactor_grid_search(self):
-        # Assert unique values are correct
-        assert (set(CA_vals).issuperset(set([1, 3, 5]))) and (
-            set(T_vals).issuperset(set([300, 500, 700]))
+        # self.assertTrue(unique values are correct
```
Looks like this got changed by mistake with a find-and-replace

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -416,13 +391,14 @@ def test_reactor_grid_search_bad_model(self):
-        # Assert length is correct
-        assert (len(CA_vals) == 9) and (len(T_vals) == 9)
-        assert (len(set(CA_vals)) == 3) and (len(set(T_vals)) == 3)
+        # self.assertTrue(length is correct
```
Looks like this got changed by mistake with a find-and-replace

mrmundt commented on pyomo/contrib/doe/tests/test_doe_build.py:
```diff
@@ -0,0 +1,478 @@
+
+ipopt_available = SolverFactory("ipopt").available()
+
+DATA_DIR = Path(__file__).parent
+file_path = DATA_DIR / ".." / "examples" / "result.json"
```
In case it's helpful - we have a nifty utility called "this_file_dir":
```
from pyomo.common.fileutils import this_file_dir
currdir = this_file_dir()
```

mrmundt replied:
Additional thought: it is better practice to us `os.path.sep` and `os.path.join`

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -416,13 +391,14 @@ def test_reactor_grid_search_bad_model(self):
-        # Assert unique values are correct
-        assert (set(CA_vals).issuperset(set([1, 3, 5]))) and (
-            set(T_vals).issuperset(set([300, 500, 700]))
+        # self.assertTrue(unique values are correct
```
Looks like this got changed by mistake with a find-and-replace

mrmundt commented on pyomo/contrib/doe/tests/test_doe_build.py:
```diff
@@ -0,0 +1,478 @@
+
+        count = 0
+        for k, v in doe_obj.compute_FIM_model.experiment_outputs.items():
+            self.assertTrue(pyo.value(k) == stuff[count])
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_build.py:
```diff
@@ -0,0 +1,478 @@
+
+        count = 0
+        for k, v in doe_obj.compute_FIM_model.experiment_inputs.items():
+            self.assertTrue(pyo.value(k) == stuff[count])
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_build.py:
```diff
@@ -0,0 +1,478 @@
+
+        count = 0
+        for k, v in doe_obj.compute_FIM_model.measurement_error.items():
+            self.assertTrue(pyo.value(k) == stuff[count])
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_build.py:
```diff
@@ -0,0 +1,478 @@
+
+        count = 0
+        for k, v in doe_obj.compute_FIM_model.unknown_parameters.items():
+            self.assertTrue(pyo.value(k) == stuff[count])
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -0,0 +1,406 @@
+        doe_obj.run_doe()
+
+        # self.assertTrue(model solves
+        self.assertTrue(doe_obj.results["Solver Status"] == "ok")
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -0,0 +1,406 @@
+
+        doe_obj.run_doe()
+
+        self.assertTrue(doe_obj.results["Solver Status"] == "ok")
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -0,0 +1,406 @@
+
+        doe_obj.run_doe()
+
+        self.assertTrue(doe_obj.results["Solver Status"] == "ok")
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -0,0 +1,406 @@
+
+        doe_obj.run_doe()
+
+        self.assertTrue(doe_obj.results['Solver Status'] == "ok")
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -0,0 +1,406 @@
+
+        doe_obj.run_doe()
+
+        self.assertTrue(doe_obj.results["Solver Status"] == "ok")
```
Could use `self.assertEqual` instead here.

mrmundt commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -0,0 +1,406 @@
+
+        doe_obj = DesignOfExperiments(**DoE_args)
+
+        doe_obj.compute_FIM(method="sequential")
+
+    def test_compute_FIM_seq_forward(self):
+        fd_method = "forward"
+        obj_used = "determinant"
+
+        experiment = FullReactorExperiment(data_ex, 10, 3)
```
There are no assertions on either of these - is that intentional?

djlaky replied:
This is intentional. The idea is to make sure the problem solves. I can add a check to make sure the output is what we expect, but that will take me a bit of time.

The four tests that solve a problem and have no assertions will become more "test-like" when we add the safeguarding for model solving noted with the immediate TODO's in #3345.

mrmundt replied:
That's fine - not all tests need asserts. It might be worth putting a comment summarizing this comment, however:
"These tests ensure that the options run without errors."

mrmundt commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -0,0 +1,406 @@
+        experiment = FullReactorExperiment(data_ex, 10, 3)
+
+        DoE_args = get_standard_args(experiment, fd_method, obj_used)
+
+        doe_obj = DesignOfExperiments(**DoE_args)
+
+        doe_obj.compute_FIM(method="kaug")
+
+    def test_compute_FIM_seq_backward(self):
+        fd_method = "backward"
+        obj_used = "determinant"
+
+        experiment = FullReactorExperiment(data_ex, 10, 3)
+
```
There are no asserts on either of these - is that intentional?

mrmundt commented on pyomo/contrib/doe/doe.py:
```diff
@@ -25,1023 +25,749 @@
-            If True, only the lower triangle of the FIM is computed. Default is True.
         """
+        # Deprecation error
+        if 'create_model' in kwargs:
+            raise ValueError(
+                "DEPRECATION ERROR: Pyomo.DoE has been refactored. The current interface utilizes Experiment "
+                "objects that label unknown parameters, experiment inputs, experiment outputs and measurement "
+                "error. This avoids string-based naming which is fragile. For instruction to use the new "
+                "interface, please see the Pyomo.DoE under the contributed packages documentation at "
+                "`https://pyomo.readthedocs.io/en/latest/contributed_packages/doe/doe.html`"
+            )
```
There is a utility for this:
```
from pyomo.common.deprecation import deprecation_warning
deprecation_warning(message, version='6.7.4.dev0')
```

djlaky replied:
Does the warning cause the code to stop? Or do I need to include an additional error?

mrmundt replied:
You would need to include a `raise` statement if the idea is to get the code to stop at this point, yes.

mrmundt commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1070,127 +792,126 @@ def initialize_jac(m, i, j):
-            p: parameter
-            n: response
+            n: experimental output
+            p: unknown parameter
```
I have generally negative feelings about one-letter var names unless it's for a short loop (e.g., `for i in range(whatever)`). It lowers readability.

djlaky replied:
We're inclined to keep it the way it is since the function is for developers. However, I'd be happy to change it if it's required for the merge.

blnicho commented on pyomo/contrib/doe/__init__.py:
```diff
@@ -19,7 +19,8 @@
     "instructions to use the new interface, please see the Pyomo.DoE under the contributed "
-    "packages documentation at `https://pyomo.readthedocs.io/en/latest/contributed_packages/doe/doe.html`"
+    "packages documentation at "
+    #    "`https://pyomo.readthedocs.io/en/latest/contributed_packages/doe/doe.html`"
```
@djlaky you should be able to revert this change. The URL checker was failing from a URL unrelated to this PR that was temporarily unavailable but seems to be back online now.

blnicho reviewed (approved):
Assuming the Jenkins tests pass, I think this is ready

djlaky commented:
> Assuming the Jenkins tests pass, I think this is ready

@blnicho I added a single line to the code today to help with initialization of the models. It will also be under the "To-Dos" to safegaurd the solver call against bad solves for #3345.

mrmundt commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1302,133 +1220,1012 @@ def det_general(m):
+            outputs = [k.name for k, v in model.measurement_error.items()]
+        except:
+            raise RuntimeError(
+                "Experiment model does not have suffix " + '"measurement_error".'
```
This is weird?

mrmundt commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1302,133 +1220,1012 @@ def det_general(m):
+        plt.pyplot.rc("ytick", labelsize=font_tick)
+        ax = fig.add_subplot(111)
+        params = {"mathtext.default": "regular"}
+        # plt.rcParams.update(params)
```
Leftover?

mrmundt commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1302,133 +1220,1012 @@ def det_general(m):
+        plt.pyplot.rc("ytick", labelsize=font_tick)
+        ax = fig.add_subplot(111)
+        params = {"mathtext.default": "regular"}
+        # plt.rcParams.update(params)
```
Leftover?

mrmundt reviewed (approved):
There are a couple of lines that I think need to be removed, but they aren't essential, so approve.
